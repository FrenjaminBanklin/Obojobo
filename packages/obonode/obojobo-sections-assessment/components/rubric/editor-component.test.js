import React from 'react'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'
import Rubric from './editor-component'
import Dispatcher from 'obojobo-document-engine/src/scripts/common/flux/dispatcher'

import ModalUtil from 'obojobo-document-engine/src/scripts/common/util/modal-util'
jest.mock('obojobo-document-engine/src/scripts/common/util/modal-util')
import { Transforms } from 'slate'
import { ReactEditor } from 'slate-react'
jest.mock('slate-react')
jest.mock(
	'obojobo-document-engine/src/scripts/oboeditor/components/node/with-slate-wrapper',
	() => item => item
)

const RUBRIC_NODE = 'ObojoboDraft.Sections.Assessment.Rubric'
const ASSESSMENT_NODE = 'ObojoboDraft.Sections.Assessment'

describe('Rubric editor', () => {
	beforeEach(() => {
		jest.resetAllMocks()
	})

	test('Rubric renders', () => {
		const component = renderer.create(
			<Rubric
				element={{
					content: {
						mods: [
							{ reward: 3, attemptCondition: '$last_attempt' },
							{ reward: 3, attemptCondition: '1' },
							{ reward: -3, attemptCondition: '[1,$last_attempt' },
							{ reward: -3, attemptCondition: '[$last_attempt,5]' }
						]
					}
				}}
				editor={{ selection: [0, 0] }}
			/>
		)
		const tree = component.toJSON()

		expect(tree).toMatchSnapshot()
	})

	test('Rubric calls updateNodeFromState when saved', () => {
		jest.spyOn(Transforms, 'setNodes').mockReturnValue(true)
		renderer.create(
			<Rubric
				element={{
					content: { unableToPassType: 'set-value', mods: [] }
				}}
				editor={{ selection: [0, 0] }}
			/>
		)

		Dispatcher.trigger('editor:save')
		expect(Transforms.setNodes).toHaveBeenCalled()
	})

	test('Rubric calls Dispatcher.off when unmount', () => {
		jest.spyOn(Dispatcher, 'off').mockReturnValue(true)
		const component = renderer.create(
			<Rubric
				element={{
					content: { unableToPassType: 'set-value', mods: [] }
				}}
			/>
		)
		component.getInstance().componentWillUnmount()
		expect(Dispatcher.off).toHaveBeenCalled()
	})

	test('Rubric changes type', () => {
		jest.spyOn(Transforms, 'setNodes').mockReturnValue(true)
		const component = mount(
			<Rubric
				element={{
					content: { mods: [] }
				}}
			/>
		)
		component
			.find('input')
			.at(0)
			.simulate('click', { stopPropagation: jest.fn() })

		component
			.find('input')
			.at(1)
			.simulate('click', { stopPropagation: jest.fn() })
		component
			.find('input')
			.at(1)
			.simulate('change', { target: { value: 'pass-fail' } })

		expect(Transforms.setNodes).toHaveBeenCalled()
	})

	test('Rubric changes pass score', () => {
		jest.spyOn(Transforms, 'setNodes').mockReturnValue(true)
		const component = mount(
			<Rubric
				element={{
					content: { mods: [] }
				}}
				editor={{
					toggleEditable: jest.fn()
				}}
			/>
		)

		component
			.find('input')
			.at(2)
			.simulate('focus')
		component
			.find('input')
			.at(2)
			.simulate('click', { stopPropagation: jest.fn() })
		component
			.find('input')
			.at(2)
			.simulate('change', { target: { name: 'passingAttemptScore', value: 100 } })
		component
			.find('input')
			.at(2)
			.simulate('blur')

		expect(component.instance().state.passingAttemptScore).toBe(100)
	})

	test('Rubric updates slate state', () => {
		jest.spyOn(Transforms, 'setNodes').mockReturnValue(true)
		const component = mount(
			<Rubric
				element={{
					content: { unableToPassType: 'set-value', mods: [] }
				}}
			/>
		)

		component.instance().updateNodeFromState()

		expect(Transforms.setNodes).toHaveBeenCalledTimes(1)
	})

	test('Rubric changes unable to pass result', () => {
		jest.spyOn(Transforms, 'setNodes').mockReturnValue(true)
		const component = mount(
			<Rubric
				element={{
					content: { unableToPassType: 'set-value', mods: [] }
				}}
			/>
		)

		component
			.find('select')
			.at(2)
			.simulate('click', { stopPropagation: jest.fn() })
		component
			.find('select')
			.at(2)
			.simulate('change', { target: { value: '$highest_attempt_score' } })

		component
			.find('input')
			.at(5)
			.simulate('click', { stopPropagation: jest.fn() })
		component
			.find('input')
			.at(5)
			.simulate('change', { target: { name: 'unableToPassResult', value: '100' } })

		expect(component.instance().state.unableToPassResult).toBe('100')
	})

	test('Rubric opens mod dialog', () => {
		const component = mount(
			<Rubric
				element={{
					content: { unableToPassType: 'set-value', mods: [] }
				}}
				editor={{
					children: [
						{
							type: ASSESSMENT_NODE,
							content: { attempts: 3 },
							children: [
								{
									type: RUBRIC_NODE,
									content: { unableToPassType: 'set-value', mods: [] },
									children: []
								}
							]
						}
					]
				}}
			/>
		)

		ReactEditor.findPath.mockReturnValueOnce([0, 0])

		component
			.find('button')
			.at(0)
			.simulate('click')

		expect(ModalUtil.show).toHaveBeenCalled()
	})

	test('Rubric changes mods', () => {
		jest.spyOn(Transforms, 'setNodes').mockReturnValue(true)
		const component = mount(
			<Rubric
				element={{
					content: { unableToPassType: 'set-value', mods: [] }
				}}
			/>
		)

		component.instance().changeMods({ mods: [] })

		expect(Transforms.setNodes).toHaveBeenCalled()
	})
})
