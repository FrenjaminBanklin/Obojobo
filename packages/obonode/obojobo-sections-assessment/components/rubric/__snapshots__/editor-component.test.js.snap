// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rubric editor Rubric renders 1`] = `
<div
  className="rubric pad is-type-undefined"
>
  <h2
    contentEditable={false}
  >
    Assessment Scoring
  </h2>
  <p>
    The assessment score is the final counted score for this module, and will be sent to any connected gradebook
  </p>
  <fieldset
    className="assessment-score"
  >
    <legend>
      How do you want to determine the assessment score?
    </legend>
    <label>
      <input
        checked={false}
        name="score-type"
        onChange={[Function]}
        onClick={[Function]}
        type="radio"
        value="highest"
      />
      Use the highest attempt score
    </label>
    <label>
      <input
        checked={false}
        name="score-type"
        onChange={[Function]}
        onClick={[Function]}
        type="radio"
        value="pass-fail"
      />
      Calculate based on a threshold (pass/fail)
    </label>
  </fieldset>
  <fieldset
    className="pass-fail"
  >
    <legend>
      Pass & Fail Rules
    </legend>
    <p>
      In this mode, students must achieve a certain threshold on an attempt to pass. The assessment scores will be set based on whether the student passes or fails
    </p>
    <div>
      <label>
        To 
        <b>
          pass
        </b>
        , students must achieve an attempt score of at least
        <input
          max="100"
          min="0"
          onChange={[Function]}
          onClick={[Function]}
          type="number"
        />
        %
      </label>
    </div>
    <div>
      <label>
        When 
        <b>
          passing
        </b>
        , set the recorded score to
        <select
          onChange={[Function]}
          onClick={[Function]}
        >
          <option
            value="$attempt-score"
          >
            The highest attempt score
          </option>
          <option
            value="set-value"
          >
            Specified value
          </option>
        </select>
      </label>
      <label
        className=" is-not-enabled"
      >
        <input
          disabled={true}
          max="100"
          min="0"
          onChange={[Function]}
          onClick={[Function]}
          type="number"
        />
        %
      </label>
    </div>
    <div>
      <label>
        When 
        <b>
          failing
        </b>
        ,
        <select
          onChange={[Function]}
          onClick={[Function]}
        >
          <option
            value="$attempt-score"
          >
            Set the score to the highest attempt score
          </option>
          <option
            value="no-score"
          >
            Dont set the score (no score will be sent to the gradebook)
          </option>
          <option
            value="set-value"
          >
            Set the assessment score to specified value
          </option>
        </select>
      </label>
      <label
        className=" is-not-enabled"
      >
        <input
          disabled={true}
          max="100"
          min="0"
          onChange={[Function]}
          onClick={[Function]}
          type="number"
        />
        %
      </label>
    </div>
    <div>
      <label>
        And if the student is 
        <b>
          out of attempts and still did not pass
        </b>
        ,
        <select
          onChange={[Function]}
          onClick={[Function]}
        >
          <option
            value="no-value"
          >
            Dont do anything, the failing rule will still apply
          </option>
          <option
            value="$attempt-score"
          >
            Set the score to the highest attempt score
          </option>
          <option
            value="no-score"
          >
            Dont set the score (no score will be sent to the gradebook)
          </option>
          <option
            value="set-value"
          >
            Set the assessment score to specified value
          </option>
        </select>
      </label>
      <label
        className=" is-not-enabled"
      >
        <input
          disabled={true}
          max="100"
          min="0"
          onChange={[Function]}
          onClick={[Function]}
          type="number"
        />
        %
      </label>
    </div>
  </fieldset>
  <div
    className="mods"
  >
    <div
      className="title"
    >
      Extra Credit & Penalties
    </div>
    <div
      className="obojobo-draft--components--button is-not-dangerous align-center"
    >
      <button
        className="button"
        contentEditable={false}
        onClick={[Function]}
      >
        Edit...
      </button>
    </div>
    <ul>
      <li>
        <b>
          <span
            className="reward"
          >
            Add
          </span>
           
          3
          %
        </b>
        <span>
           If a student passes on 
          the last attempt
           
        </span>
      </li>
      <li>
        <b>
          <span
            className="reward"
          >
            Add
          </span>
           
          3
          %
        </b>
        <span>
           If a student passes on 
          attempt 1
           
        </span>
      </li>
      <li>
        <b>
          <span
            className="deduct"
          >
            Deduct
          </span>
           
          3
          %
        </b>
        <span>
           
          If a student passes on attempt 
          1
           through 
          the last attempt
           
        </span>
      </li>
      <li>
        <b>
          <span
            className="deduct"
          >
            Deduct
          </span>
           
          3
          %
        </b>
        <span>
           
          If a student passes on attempt 
          the last attempt
           through 
          5
           
        </span>
      </li>
    </ul>
  </div>
</div>
`;
